hail/python/hailtop/utils/utils.py

```diff
@@ -158,8 +170,54 @@ def __init__(self, worker_pool):
         self._waiting = True
         if self._n_complete == self._n_submitted:
             self._done.set()
+        if not self._ignore_errors and self._errors:
+            self._done.set()
+
         await self._done.wait()
 
+        if self._errors:
+            raise self._errors[0]
+
+
+class WaitableBunch:
+    def __init__(self, shared_pool):
+        self.shared_pool = shared_pool
+        self._errors = []
+        self._done = asyncio.Event()
+        self._waiting = False
+        self._n_submitted = 0
+        self._n_complete = 0
+
+    async def call(self, f, *args, **kwargs):
+        assert not self._waiting
+        self._n_submitted += 1
+
+        async def invoke():
+            try:
+                await f(*args, **kwargs)
+            except Exception as err:  # pylint: disable=broad-except
+                self._errors.append(err)
+                self._done.set()
+            finally:
+                self._n_complete += 1
+                if self._waiting and (self._n_complete == self._n_submitted):
+                    self._done.set()
+
+        await self.shared_pool._worker_pool.call(invoke)
+
+    async def wait(self):
+        assert not self._waiting
+        self._waiting = True
+        if self._n_complete == self._n_submitted:
+            self._done.set()
+        if self._errors:
+            self._done.set()
```

Comment:
It seems simpler to remove `_waiting`, and just always set self._done when we reach all tasks complete, then here we needn't have either check. If it already happened, then done will be set. Otherwise, it will be set while we're waiting on it.

Reply:
I don't think you can do this.

Reply:
Can you explain what I misunderstand? It seems like the timeline of an object like this is:
1. get created
2. add some jobs via `call`
3. `wait` once on this bunch

If that understanding is true, then `done` starts out unset and only transitions to `set` if the bunch is finished (due either to an error or to all jobs finishing). If we unconditionally set it when there's an error or when all jobs finish, then when we enter wait either it is set or it will be set in the future (assuming the jobs terminate). There's no need to check any conditions in `wait`.

Reply:
See the above comment.